### PR TITLE
feat: support dtm upgrade

### PR DESCRIPTION
**Pull request by hxcGit**
Signed-off-by: hxcGit <houxc_mail@163.com>

## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [ ] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [ ] My code has the necessary comments and documentation (if needed).
- [ ] I have added relevant tests

## Description

support dtm upgrade
### usage

```shell
dtm upgrade
```

## Related Issues

#674 

### Test in MacOS
- Dtm version was manually modified before test
image

### Test in Linux/amd64
image

### Do not support to upgrade in development version
image

### CODE REVIEW

File: cmd/devstream/upgrade.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+	"github.com/spf13/cobra"
+
+	"github.com/devstream-io/devstream/internal/pkg/upgrade"
+)
+
+var upgradeCMD = &cobra.Command{
+	Use:   "upgrade",
+	Short: "Upgrade dtm to the latest version",
+	Long:  `Upgrade dtm to the latest version.`,
```
Comment (daniel-hutao) [inline cmd/devstream/upgrade.go:14]: latest release version

File: internal/pkg/upgrade/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/Masterminds/semver"
+	"github.com/tcnksm/go-input"
+
+	"github.com/devstream-io/devstream/pkg/util/log"
```
Comment (daniel-hutao) [inline internal/pkg/upgrade/upgrade.go:12]: Ensure correct formatting, same as above.

File: internal/pkg/upgrade/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	return oldSemVer, newSemVer, nil
+}
+
+// applyUpgrade use os.Rename replace old version dtm with the latest one
+// (1) rename `dtm` to `dtm-bak`.
+// (2) rename `dtm-tmp` to `dtm`.
+// (3) grant new `dtm` execute permission.
+// (4) remove `dtm-bak` binary file.
```
Comment (daniel-hutao) [inline internal/pkg/upgrade/upgrade.go:127]: good comments

**Comment by iyear** [comment]
I think the `semver` pkg can cover more version situations. Can you add a commit to refactor this code? Thanks!

https://github.com/devstream-io/devstream/blob/7edf2749cf3f64c90ccfed626fd58cae6376f043/internal/pkg/version/version.go#L13-L16

File: cmd/devstream/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package main
+
+import (
+	"github.com/devstream-io/devstream/pkg/util/log"
+
+	"github.com/spf13/cobra"
+
+	"github.com/devstream-io/devstream/internal/pkg/upgrade"
```
[the remark this answers is not in the document]
Comment (hxcGit) [reply cmd/devstream/upgrade.go:6]: Done

File: cmd/devstream/upgrade.go (hunk 1 above)
Comment (hxcGit) [reply cmd/devstream/upgrade.go]: Done

File: internal/pkg/upgrade/upgrade.go (hunk 2 above)
Comment (hxcGit) [reply internal/pkg/upgrade/upgrade.go]: Done

**Comment by hxcGit** [comment]
@daniel-hutao @IronCore864 @steinliber  PTAL

**Comment by hxcGit** [comment]
@iyear  I'll check it out.

**Comment by iyear** [comment]
> @iyear I'll check it out.

Sorry, I just deleted the comment, I think it would be more appropriate to open an new issue to explain the situation.😂

File: internal/pkg/upgrade/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	dtmBakFilePath := filepath.Join(workDir, dtmBakFileName)
+	dtmTmpFilePath := filepath.Join(workDir, dtmTmpFileName)
+
+	if err := os.Rename(dtmFilePath, dtmBakFilePath); err != nil {
+		return err
+	}
+	log.Debug("Dtm upgrade: rename dtm to dtm-bak successfully.")
+
+	if err := os.Rename(dtmTmpFilePath, dtmFilePath); err != nil {
+		return err
```
Comment (steinliber) [inline internal/pkg/upgrade/upgrade.go:139]: if this failed, you should rollback changes before

File: pkg/util/github/release.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/devstream-io/devstream/pkg/util/log"
+)
+
+func (c *Client) LatestReleaseTagName() (string, error) {
```
Comment (steinliber) [inline pkg/util/github/release.go:11]: this func is a little incomprehensible  by name, you can add a comment to describe this func

**Review by steinliber — COMMENTED** [review]
func works great

File: pkg/util/github/release.go (hunk 6 above)
Comment (hxcGit) [reply pkg/util/github/release.go:11]: How about `GetLatestReleaseTagName`?

File: internal/pkg/upgrade/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	log.Debugf("Dtm upgrade: work path is : %v", workDir)
+
+	// 1. Get the latest release version
+	ghOptions := &github.Option{
+		Org:      dtmOrg,
+		Repo:     dtmRepo,
+		NeedAuth: false,
+		WorkPath: workDir,
+	}
+
```
[the remark this answers is not in the document]
Comment (hxcGit) [reply internal/pkg/upgrade/upgrade.go:45]: Is there any benefit to doing this? Do you mean to add a new function that returns latest release tagName?

[the remark this answers is not in the document]
Comment (steinliber) [reply internal/pkg/upgrade/upgrade.go:45]: I mean you can put create github option and init github client in separate func, beacuse the creation of new client logic is not very related to upgrade func

File: pkg/util/github/release.go (hunk 6 above)
Comment (steinliber) [reply pkg/util/github/release.go:11]: that's great and more clear

**Review by steinliber — APPROVED** [review]

**Pull request merged**